x-pack/plugins/upgrade_assistant/server/lib/reindexing/worker.ts

```diff
@@ -0,0 +1,181 @@
+ *   (1) A longer loop that polls for reindex operations that are in progress. If any are found, loop (2) is started.
+ *   (2) A tighter loop that pushes each in progress reindex operation through ReindexService.processNextStep. If all
+ *       updated reindex operations are complete, this loop will terminate.
+ *
+ * The worker can also be forced to start loop (2) by calling forceRefresh(). This is done when we know a new reindex
+ * operation has been started.
+ *
+ * This worker can be ran on multiple nodes without conflicts or dropped jobs. Reindex operations are locked by the
+ * ReindexService and if any operation is locked longer than the ReindexService's timeout, it is assumed to have been
+ * locked by a node that is no longer running (crashed or shutdown). In this case, another node may safely acquire
+ * the lock for this reindex operation.
+ */
+export class ReindexWorker {
+  private static workerSingleton?: ReindexWorker;
+  private continuePolling: boolean = false;
+  private updateOperationLoopRunning: boolean = false;
+  private inProgressOps: ReindexSavedObject[] = [];
+  private readonly reindexService: ReindexService;
+
+  constructor(
+    private client: SavedObjectsClient,
+    private credentialStore: CredentialStore,
+    private callWithRequest: CallClusterWithRequest,
+    private callWithInternalUser: CallCluster,
+    private readonly log: Server['log']
+  ) {
+    if (ReindexWorker.workerSingleton) {
+      throw new Error(`More than one ReindexWorker cannot be created.`);
+    }
+
+    this.reindexService = reindexServiceFactory(
+      this.callWithInternalUser,
+      reindexActionsFactory(this.client, this.callWithInternalUser)
+    );
+
+    ReindexWorker.workerSingleton = this;
+  }
+
+  /**
+   * Begins loop (1) to begin checking for in progress reindex operations.
+   */
+  public start = () => {
+    this.log(['debug', ...LOG_TAGS], `Starting worker...`);
+    this.continuePolling = true;
+    this.pollForOperations();
+  };
+
+  /**
+   * Stops the worker from processing any further reindex operations.
+   */
+  public stop = () => {
+    this.log(['debug', ...LOG_TAGS], `Stopping worker...`);
+    this.updateOperationLoopRunning = false;
+    this.continuePolling = false;
+  };
+
+  /**
+   * Should be called immediately after this server has started a new reindex operation.
+   */
+  public forceRefresh = () => {
+    this.refresh();
+  };
+
+  /**
+   * Returns whether or not the given ReindexOperation is in the worker's queue.
+   */
+  public includes = (reindexOp: ReindexSavedObject) => {
+    return this.inProgressOps.map(o => o.id).includes(reindexOp.id);
+  };
+
+  /**
+   * Runs an async loop until all inProgress jobs are complete or failed.
+   */
+  private startUpdateOperationLoop = async () => {
+    this.updateOperationLoopRunning = true;
+
+    while (this.inProgressOps.length > 0) {
+      this.log(['debug', ...LOG_TAGS], `Updating ${this.inProgressOps.length} reindex operations`);
```

Comment:
We might want to add a delay to this.

```
  log   [18:12:24.208] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.208] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.209] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.210] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.211] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.211] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.212] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.213] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.214] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.214] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.215] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
  log   [18:12:24.216] [debug][reindex_worker][upgrade_assistant] Updating 1 reindex operations
```

Reply:
Do you think we should slow down the logging? I don't think we should delay the processing loop since most state changes are very quick. The only state change that is "slow" is waiting for the reindex task to complete (which is essentially a no-op).

We could add a delay if all the inProgressOps are currently waiting for the reindex task to complete so we don't poll ES in a tight loop.

Reply:
is there no delay?  it looks like it's running every millisecond?

Reply:
Talked to @tylersmalley about this, it looks like this happens when ES goes down and this turns into a very fast tight loop. I'll add some handling for this.